### PR TITLE
Updates to V4 protocol support

### PR DESCRIPTION
**Pull request by nevans**
Recreating #101

### CODE REVIEW

File: lib/staccato/V4/event.rb
Comment (tpitale) [inline lib/staccato/V4/event.rb:1]: I _thought_ I had already moved this 🤔

File: lib/staccato/adapter/logger.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@ module Adapter
     class Logger # The Ruby HTTP Library Adapter
       DEFAULT_FORMATTER = lambda {|params| params.map {|k,v| [k,v].join('=')}.join(' ')}
 
+      attr_reader :uri, :logger
+      attr_accessor :formatter
```
Comment (tpitale) [inline lib/staccato/adapter/logger.rb:9]: I think I'd prefer not to expose this. It should be passed in on `new` and not really changed later. If someone requests this down the line we can reconsider.

File: lib/staccato/adapter/net_http.rb [hunk 2]
```diff
@@ -3,7 +3,11 @@
 module Staccato
   module Adapter
     module Net
-      class HTTP # The net/http Standard Library Adapter
+      # The net/http Standard Library Adapter
+      class HTTP
+
+        attr_reader :uri
```
Comment (tpitale) [inline lib/staccato/adapter/net_http.rb:9]: Is this reader used outside the adapter?

File: lib/staccato/v4.rb [hunk 3]
```diff
@@ -1,58 +1,56 @@
...
     #
-    # @param measurement_id [String] the required id provided by google, i.e., `G-XXXXXXXX`
-    # @param client_id [String, Integer, nil] a unique id to track the session of
-    #   an individual user
-    # @param options [Hash] options for use in all hits from this tracker
     # @yield [Staccato::Tracker] the new tracker
     # @return [Staccato::Tracker] a new tracker is returned
-    def self.tracker(measurement_id, api_secret, client_id = nil, options = {})
-      klass = measurement_id.nil? ? Staccato::V4::NoopTracker : Staccato::V4::Tracker
+    def self.tracker(measurement_id, api_secret, client_id, **options)
```
Comment (tpitale) [inline lib/staccato/v4.rb:25]: Can you share the benefit to doing `**options` here? Is it because you've got explicit named args in the tracker class?

File: lib/staccato/V4/event.rb
Comment (nevans) [reply lib/staccato/V4/event.rb:1]: See #103.  :)

File: lib/staccato/adapter/logger.rb (hunk 1 above)
Comment (nevans) [reply lib/staccato/adapter/logger.rb:9]: I generally agree.  I _think_ I kept it r/w simply because stdlib's `Logger#formatter` is an `attr_accessor`.  And it (very slightly) simplified the spec setup.

File: lib/staccato/v4/event.rb [hunk 4]
```diff
@@ -1,76 +1,145 @@
...
+      self.class.recommended_params.map {|field|
         [field, options[field]] unless options[field].nil?
       }.compact.to_h
     end
 
-    # include BooleanHelpers
-    # adds convert_booleans()
+    # The custom event parameters to be sent for this event
+    # @return (see params)
+    def custom_params
```
Comment (tpitale) [inline lib/staccato/v4/event.rb:105]: This doesn't appear to be used yet?

File: lib/staccato/v4/events/share.rb [hunk 5]
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Comment (tpitale) [inline lib/staccato/v4/events/share.rb:1]: It's wild to me that ruby still requires this 😆

File: lib/staccato/adapter/net_http.rb (hunk 2 above)
Comment (nevans) [reply lib/staccato/adapter/net_http.rb:9]: I used it in the specs to verify collection vs validation URIs.  And I like to keep `attr_reader`s for this sort of thing for more "ergonomic" debug access.

File: lib/staccato/v4/events/item.rb [hunk 6]
```diff
@@ -18,7 +20,7 @@ module Staccato::V4
     :item_variant,
     :location_id,
     :price,
-    :quantity
+    :quantity,
+    keyword_init: true
```
Comment (tpitale) [inline lib/staccato/v4/events/item.rb:24]: That's a cool addition. 👍🏼

File: lib/staccato/adapter/logger.rb (hunk 1 above)
Comment (tpitale) [reply lib/staccato/adapter/logger.rb:9]: 👍🏼

File: lib/staccato/adapter/net_http.rb (hunk 2 above)
Comment (tpitale) [reply lib/staccato/adapter/net_http.rb:9]: If it is, should the `Validate` adapter also have a uri reader?

Comment (tpitale) [reply lib/staccato/adapter/net_http.rb:9]: Nevermind, it delegates to the adapter.

**Review by tpitale — CHANGES_REQUESTED** [review]
Left some comments, have to comment here for them to not be pending? 🙃

File: lib/staccato/v4/event.rb [hunk 7]
```diff
@@ -1,76 +1,145 @@
...
+          .split('::')
+          .last
+          .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
+          .gsub(/([a-z\d])([A-Z])/, '\1_\2')
+          .tr('-', '_')
+          .downcase
+          .to_sym
+      end
+
+      def param(name)
```
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4/event.rb:63]: If we want to have a DSL … maybe `field` AND `fields`? So we could replace the `FIELDS`?

File: lib/staccato/v4/event.rb [hunk 8]
```diff
@@ -1,76 +1,145 @@
...
     def as_json
-      {
-        name: name,
-        params: event_params
-      }
+      Staccato::V4.encode_params_json({name: name, params: params})
     end
 
-    private
+    def ==(other) other.class == self.class && other.as_json == as_json end
```
Comment (tpitale) [inline lib/staccato/v4/event.rb:84]: Is this used? I'm not seeing it, but I could be missing it.

File: lib/staccato/v4.rb (hunk 3 above)
Comment (nevans) [reply lib/staccato/v4.rb:25]: Using `**kwargs` vs `options = {}` behaves differently in ruby 2.0-2.6 vs 3.0+ (with 2.7 splitting the difference).  It's is backward's compatible to use faux-kwargs when calling a method that uses `options = {}` in its definition.  The difference is _especially_ noticeable when you precede the kwargs with optional params, but I changed `client_id` to be required.

So, I think the main reason for making this change right _here_ is that `**options` is a better semantic match.  When I see kwargs syntax in rdoc (or yard) it _explicitly_ says to me "this gobbles up all remaining keyword arguments (and probably forwards)" vs "the fourth positional argument is an options object, probably a hash"

File: lib/staccato/v4.rb [hunk 9]
```diff
@@ -1,58 +1,56 @@
...
-    def self.ga_collection_uri
-      URI('https://www.google-analytics.com/mp/collect')
-    end
-
-    # The tracking endpoint we use to submit requests to GA
-    def self.validation_uri
-      URI('https://www.google-analytics.com/debug/mp/collect')
+    # @return [Integer] the number of microseconds since the UNIX epoch
+    def self.timestamp_micros(time = Time.now)
+      time = Time.now if time == true || time.nil?
```
[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4.rb:36]: It's microseconds.  Thus the name.  :)

[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4.rb:36]: It's in the docs somewhere... but the docs can be confusing.

File: lib/staccato/v4/event.rb [hunk 10]
```diff
@@ -1,76 +1,145 @@
...
 
-      events[event.event_name] = event
     end
 
-    def initialize(tracker, options = {})
-      self.tracker = tracker
-      self.options = Staccato::OptionSet.new(options)
+    module ClassMethods
+
+      attr_reader :recommended_params
```
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4/event.rb:49]: Oh … this is in `ClassMethods` … the other is the instance.

File: lib/staccato/v4/event.rb (hunk 4 above)
Comment (nevans) [reply lib/staccato/v4/event.rb:105]: It's for the end-user to use as they see fit, and for documentation.  All event params (both "recommended" and "custom") are stored on `params` and encoded by `as_json`.

File: lib/staccato/v4/event.rb [hunk 11]
```diff
@@ -1,76 +1,145 @@
...
+    # The custom event parameters to be sent for this event
+    # @return (see params)
+    def custom_params
+      official = self.class.recommended_params
+      params.delete_if {|k, v| official.include? k }
+    end
+
+    # Adds and updates event parameters with the values in the params argument.
+    # @return self
+    def update(params)
```
Comment (tpitale) [inline lib/staccato/v4/event.rb:112]: Also maybe not used?

File: lib/staccato/v4/events/share.rb (hunk 5 above)
Comment (nevans) [reply lib/staccato/v4/events/share.rb:1]: Yeah, I really wish it had become the default in ruby 3.0!  Maybe in 3.3?

File: lib/staccato/v4/event.rb [hunk 12]
```diff
@@ -1,76 +1,145 @@
...
+
+    # Adds parameters to the event, where the event parameter wasn't already set.
+    #
+    # Note: This behaves like `Hash#with_defaults!` from `ActiveSupport`.
+    #
+    # @return self
+    def with_defaults!(defaults)
+      defaults.to_hash.each_pair do |key, value|
+        key = key.to_sym
+        options[key] = value unless options.to_h.key?(key)
```
Comment (tpitale) [inline lib/staccato/v4/event.rb:128]: Would `||=` work here?

If not, should `options` have a `has_key?`?

File: lib/staccato/v4/tracker.rb [hunk 13]
```diff
@@ -1,155 +1,334 @@
...
+    # @param default_event_params [Hash] default event params used by all events
+    #     sent from this tracker
+    def initialize(measurement_id, api_secret, client_id,
+                   user_id: nil, user_properties: nil,
+                   non_personalized_ads: nil,
+                   time: nil, timestamp_micros: nil,
+                   default_event_params: {})
+      raise ArgumentError, 'measurement_id is required' unless measurement_id
+      raise ArgumentError, 'api_secret is required'     unless api_secret
+      raise ArgumentError, 'client_id is required'      unless client_id
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:39]: I think letting client_id be a random UUID as it does in V3 is acceptable.

File: lib/staccato/v4/event.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4/event.rb:63]: > I suspect if we do this now, we should just switch to this and get rid of FIELDS. I'd also like to see it documented in the README in a section on custom even classes.

Yeah, I haven't documented anything in the README.  🙁

File: lib/staccato/v4/tracker.rb [hunk 14]
```diff
@@ -1,155 +1,334 @@
...
+    #
+    #    tracker.adapters << Staccato::Adapter::Logger.new
+    #
+    # @param adapter [Object] an adapter object, or use one of the following
+    #     symbols to configure a default adapter: `:net_http`, `:validation`,
+    #     `:logger`.
+    # @param logger [Logger] used when `adapter` is `:logger`
+    # @param formatter [Proc] used when `adapter` is `:logger`
+    def add_adapter(adapter, logger = nil, formatter = nil)
+      case adapter
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:79]: Accepting a symbol here is interesting. I'm not sure it's a necessary feature for this PR. I haven't had a feature request for this, and I'd like to consider it more before adding it to the API.

File: lib/staccato/v4/tracker.rb [hunk 15]
```diff
@@ -1,155 +1,334 @@
...
+    def events; @events.dup.freeze end
+
+    # Add an event to the events to be sent, and return self.
+    #
+    # @param time [Time] The time to associate with the event.  This should only
+    #     be set to record events that happened in the past.  It will be
+    #     translated into a timestamp_micros event parameter.
+    # @param params [Time] event parameters to me used with the event
+    # @return self
+    def add(event, time: nil, **params)
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:154]: These changes are my favorite part of the PR. I think this is a great addition to support "custom" events.

File: lib/staccato/v4/tracker.rb [hunk 16]
```diff
@@ -1,155 +1,334 @@
...
+        self
       end
-    end
 
-    # post the hit to GA collection endpoint
-    # @return [Net::HTTPOK] the GA api always returns 200 OK
-    def track
-      post
+      define_method :"#{event_name}!" do |**params|
+        add(event_klass, **params).then { track }
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:197]: 👍🏼

File: lib/staccato/v4/tracker.rb [hunk 17]
```diff
@@ -1,155 +1,334 @@
...
+        adapters.map {|adapter| adapter.post_with_body(query_params, body)}
+      end
     end
   end
 
   # A tracker which does no tracking
   #   Useful in testing
-  class NoopTracker
+  class NoopTracker < Tracker
     attr_accessor :hit_defaults
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:303]: I probably should have removed this as it's not a hit anymore, just an event. Oops!

File: lib/staccato/v4/event.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4/event.rb:49]: This is the `ClassMethods` half of the DSL.  It could be renamed to `recommended_param_names` to avoid confusion?

**Comment by tpitale** [comment]
Thank you again @nevans. I left a bunch of comments. Appreciate all the work. You can address the comments here, or split the PR up. I'll do my best to review and merge this.

File: lib/staccato/V4/event.rb
Comment (tpitale) [reply lib/staccato/V4/event.rb:1]: Merged!

File: lib/staccato/v4/event.rb (hunk 11 above)
Comment (nevans) [reply lib/staccato/v4/event.rb:112]: It's used by `Tracker#add`:
```ruby
        @events << event.with_defaults!(default_event_params).update(params)
```

File: lib/staccato/v4.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4.rb:36]: Yeah … there's a lot of them.

File: lib/staccato/v4/events/share.rb (hunk 5 above)
Comment (tpitale) [reply lib/staccato/v4/events/share.rb:1]: 🤞🏼

File: lib/staccato/v4/event.rb (hunk 12 above)
Comment (nevans) [reply lib/staccato/v4/event.rb:128]: I don't remember, but I might have copied this implementation from ActiveSupport.  They need to do it this way because they need to distinguish between `nil` values and no matching key.  We don't need to make that distinction, so `||=` should suffice.

File: lib/staccato/v4/event.rb (hunk 8 above)
Comment (tpitale) [reply lib/staccato/v4/event.rb:84]: Feels like `as_json` is an "expensive" method, but it's only really for the tests, cool.

File: lib/staccato/v4.rb [hunk 18]
```diff
@@ -1,58 +1,56 @@
...
-    # The default adapter to use for sending hits
-    def self.default_adapter
-      require 'staccato/adapter/net_http'
-      Staccato::Adapter::Net::HTTP
-    end
-
-    def self.validation_adapter
-      require 'staccato/adapter/validate'
-      Staccato::Adapter::Validate
+    def self.encode_params_json(params)
```
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4.rb:40]: Huh … I figured `as_json` would recursively call `as_json` when it was being encoded by the JSON library. Maybe I didn't put `as_json` on `Item`? 🤔

[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4.rb:40]: Looks like struct has as_json https://ruby-doc.org/stdlib-2.5.1/libdoc/json/rdoc/Struct.html so hopefully that would simplify things.

File: lib/staccato/v4/event.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4/event.rb:49]: Yeah … GA used to call it "fields" now events have "parameters" … so yeah, `recommended_param_names` (though I do often misspell recommended 🙃 )

File: lib/staccato/v4/tracker.rb (hunk 13 above)
Comment (nevans) [reply lib/staccato/v4/tracker.rb:39]: Yeah, I puzzled over this one.

Requiring `client_id` is a big change, but I believe it is necessary for to truly satisfy Google's v4 API.  As far as I can tell, there is no documentation anywhere for using anything other than the `gtag` or `gtm` generated `client_id`.  The big push to v4 _seems_ to be at motivated by regulatory compliance concerns.  And my interpretation of the docs is that access to standard analytics `client_id` _could_ potentially be a matter of end-user consent (depending on what you're tracking).

In other words, I believe it might be wrong for the library to automatically generate a missing `client_id`.  If users of the gem think that it is reasonable and permissible to generate a missing client_id, they should manage that themselves.

Also, using a randomly generated UUID can result in tremendously inflated session counts and your boss telling you to try something else.  Ask me how I know.  😉

File: lib/staccato/v4/event.rb (hunk 4 above)
Comment (tpitale) [reply lib/staccato/v4/event.rb:105]: I'd like to hold off on this for now until there's a use for it before we add it to the public API.

File: lib/staccato/v4.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (tpitale) [reply lib/staccato/v4.rb:36]: Found it. Yeah it is micros 🙃

File: lib/staccato/v4/tracker.rb (hunk 14 above)
Comment (nevans) [reply lib/staccato/v4/tracker.rb:79]: It really isn't necessary for this PR.

But logging is such a useful and basic feature that I wanted to remove barriers to adding it.  Also... it simplified my app's code.  ;)

File: lib/staccato/v4/tracker.rb (hunk 17 above)
Comment (nevans) [reply lib/staccato/v4/tracker.rb:303]: Yeah, I meant to delete that.  Oops.

**Comment by nevans** [comment]
@tpitale thanks for your time!  I don't know if I'll split it up this week, but there are some obvious chunks to break off as separate PRs.

File: lib/staccato/v4/tracker.rb (hunk 13 above)
Comment (tpitale) [reply lib/staccato/v4/tracker.rb:39]: I'm not sure how a random UUID would be a matter of end-user consent.

I could see it being a session issue, as it's going to make a new session for each post.

Is there a way we could split the difference? Maybe make it a configuration somewhere? `require_client_id: true` by default, and a user could disable that with false to get the random UUID back?

File: lib/staccato/v4/tracker.rb (hunk 14 above)
Comment (tpitale) [reply lib/staccato/v4/tracker.rb:79]: I agree it's a nice feature. I'm just trying to break down the PR. There's a LOT here 😄

File: lib/staccato/v4/tracker.rb [hunk 19]
```diff
@@ -1,155 +1,334 @@
...
+    #     symbols to configure a default adapter: `:net_http`, `:validation`,
+    #     `:logger`.
+    # @param logger [Logger] used when `adapter` is `:logger`
+    # @param formatter [Proc] used when `adapter` is `:logger`
+    def add_adapter(adapter, logger = nil, formatter = nil)
+      case adapter
+      when :logger
+        require 'staccato/adapter/logger'
+        uri = self.class.collection_uri
+        adapter = Staccato::Adapter::Logger.new(uri, logger, formatter)
```
Comment (tpitale) [inline lib/staccato/v4/tracker.rb:83]: Could this follow the same pattern as the other adapters in the defaults module? `self.class.logging_adapter(uri, logger, formatter)`?

**Comment by tpitale** [comment]
My laptop is dying and I need to get back to my vacation. I'll try to check back later.

File: lib/staccato/v4.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4.rb:40]: It's more that stdlib's `as_json` extensions behave very differently from ActiveSupport's `as_json`.  In particular, stdlib's `Struct#as_json` won't work for `Item`.  Also, I wanted to compact away any `nil` values.

Stdlib's `as_json` feels more appropriate for serializing data similarly to how me might YAML or Marshal.   ActiveSupport's is generally more appropriate for APIs like this one.  This allows us to not care whether either extension is loaded.

[the remark this answers is not in the document]
Comment (nevans) [reply lib/staccato/v4.rb:40]: > Maybe I didn't put as_json on Item? :thinking:

I might have been able to get away with just adding `as_json` on `Item`.  But this removes `nil` values, allows custom structs, and just seemed more generically useful.

**Comment by nevans** [comment]
> My laptop is dying and I need to get back to my vacation. I'll try to check back later.

I probably won't get to this again before Monday, so please: enjoy your vacation!

File: lib/staccato/v4/tracker.rb (hunk 13 above)
Comment (nevans) [reply lib/staccato/v4/tracker.rb:39]: > I'm not sure how a random UUID would be a matter of end-user consent.

Being forced get the real `client_id` also forces the developer into a position where it should be easy for them to also check for consent.  That _might_ encourage the developer to _consider_ whether they're allowed to send everything they send.

Not knowing the real `client_id` suggests that we haven't even attempted to check whether we have end-user consent.  We might have both on the frontend but never communicated it to the backend.  Or we might not have a `client_id` _because_ we don't have consent.

Either way, that doesn't mean we can't send _any_ data, but (as I understand it) it does mean we shouldn't send certain private or user identifying data.  It might be that your app is fine sending everything you intended to send.  In my case, we will send certain `user_properties` and custom parameters and custom events if we have consent, and we won't if we don't.

I'd prefer a config setting to an always on default.  But personally, I think it's better to force the developer to make their own decision about how to handle it:
* the above reason: force the developer to consider the consequences.
* Google's docs never give any valid alternatives to the official `client_id`.
* Unique client IDs for each tracker instance causes an _explosion_ in session count.
* The app probably has internal session identifiers of its own that are more stable.

IMO it would be fine to _document_ that `SecureRandom.uuid` seems to more-or-less work.  But, assuming you're using rails, the following would work much better and we'd be better off recommending something like it:

```ruby
class Current < ActiveSupport::CurrentAttributes
  attribute :ga_real_cid
  attribute :ga_fake_cid
  attribute :app_session_id

  def ga_cid = ga_real_cid || self.ga_fake_cid ||= app_session_id || SecureRandom.uuid

end
```

...using `before_filters` in `ApplicationController` to get and set these values appropriately.  They could even be forwarded into background jobs and used there as well, using `Current.set(ga_real_cid:, ga_fake_cid:, app_session_id:) { the bg job code }`.

File: lib/staccato/v4/tracker.rb (hunk 19 above)
Comment (nevans) [reply lib/staccato/v4/tracker.rb:83]: I considered that.  I decided against it because those were effectively accessors for shared singleton-ish objects, and this would need to generate a new logger adapter every time.

But, on reconsidering the matter, it can still be useful to delegate the decision of _how_ to generate that logger to the class.

File: lib/staccato/v4/tracker.rb (hunk 13 above)
Comment (tpitale) [reply lib/staccato/v4/tracker.rb:39]: Thanks for all the info. As you know I haven't been working with this for awhile, so this is all very useful. I appreciate the time you've taken to explain it.

Okay, as long as we document what they'll have to do in case they just need to get it working ASAP, I'm fine moving forward with this being required.

Comment (nevans) [reply lib/staccato/v4/tracker.rb:39]: @tpitale thanks!  and there's some other bits of this PR that could use more docs too.  not least of which is adding something to the README.md.  😉

File: lib/staccato/v4/event.rb (hunk 8 above)
Comment (nevans) [reply lib/staccato/v4/event.rb:84]: Re: expensive, I should probably add a short-circuit at the front of the boolean expression: `super || ...` to check object identity before checking the contents.

In this case, `as_json` is really just a way to iterate through all of the contents while also compacting away `nil`s.  And that's basically what's needed for checking value equality.

**Comment by grega** [comment]
Thank you @nevans and @tpitale for putting all of the time and effort into this! Is there anything I can help with getting it over the line (before the looming deadline of July 1st hits)?

**Comment by tpitale** [comment]
I'm happy to move forward with this if @nevans commits to helping me do README/docs updates after this 😄

**Comment by nevans** [comment]
@grega @tpitale sorry for the slow follow up.  Barring any unforseen circumstances, I'll definitely get around to cleaning up this PR _eventually_--both splitting it into smaller PRs and adding documentation.  I'll try to work on it a little bit more this week (maybe even today?), but I probably can't commit to completing it before July 1st!  (sorry)

I can say that I've been running my branch in production for for over six weeks, and I only have a few issues to report:

1. I'm using this PR's branch in my `Gemfile` with `gem "stacatto", github: "nevans/staccato", ref: "97-v4-protocol"`.  Every time I've rebased this branch, it's invalidated the commit SHA1 in my `Gemfile.lock`.  This breaks my CI and (if the `bundle install` docker layer is invalidated) my production deploys too.  But I always run `bundle update stacatto` immediately afterward anyway, so it hasn't been a huge issue.  Since I'm going to split this PR into smaller PRs anyway, I _will_ commit to no more rebasing of this branch, nor will I delete the branch until everything is merged and released.  I will merge any updates into this branch, to keep it up-to-date.  So you should be able to use this PR's branch in your `Gemfile` and I won't break it on you (any more).
2. Because Google doesn't provide any error response when you send invalid events, I've really wanted a config option to send some percentage of all events to the validation endpoint, in order to log the errors.  I've had to copy events from my logfiles and manually hit the validation endpoint to debug some invalid schema issues.  The v4 Measurement Protocol API really doesn't like it when you send numbers where it expects a string!
3. I've had some issues with Google Ads conversions not connecting with purchase events sent from the backend... but I'm not sure if that's _ever_ worked the way I had assumed it had.  It most likely broke due to another change we made to our frontend javascript at the same time as we switched to this in the backend: we switched from a bunch of different tags to `gtag.js` and then to `gtm.js`.  I really don't understand how everything connects to everything else in this Ads/Analytics/Tag Manager ecosystem, so... 🤷🏻

**Comment by tpitale** [comment]
@nevans As long as you feel you can help with docs at some point, I'll go ahead and merge this.

**Comment by nevans** [comment]
> As long as you feel you can help with docs at some point, I'll go ahead and merge this.

@tpitale Sorry, I missed this... and I never got back to finishing this otherwise.  I've been in the middle of a completely different project and it sucked away all of my attention.

This _does_ need more documentation and I should _eventually_ get to that (ideally before I completely forget what I was trying to do with this code!).  

I can't say whether merging this before or after I start on docs PR(s) is a good idea.  Breaking it into separate PRs is a good idea and it isn't a burden... to the contrary, splitting off smaller focused chunks might help the documentation project.  And I _did_ want to implement your suggestions as we discussed.  On the other hand, if you simply merged this behemoth and released a new version, and I could stop using my fork... I certainly won't complain.  ;)
